File: 0253-meeting-rooms-ii/0253-meeting-rooms-ii.cpp
class Solution {
public:
    int minMeetingRooms(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end(), [](vector<int>& a, vector<int>& b){
            if(a[0] == b[0]){
                return a[1] < b[1]; 
            }

            return a[0] < b[0]; 
        });
        
        priority_queue<int, vector<int>, greater<int>> pq; 
        pq.push(intervals[0][1]); 

        for(int i = 1; i < intervals.size(); i++){
            int startTime = intervals[i][0]; 
            int endTime = intervals[i][1]; 

            if(!pq.empty() && pq.top() <= startTime){
                pq.pop(); 
            }

            pq.push(endTime); 
        }

        return pq.size(); 
    }
};

File: 0253-meeting-rooms-ii/0253-meeting-rooms-ii.java
class Solution {
    public int minMeetingRooms(int[][] intervals) {
        Arrays.sort(intervals, (a,b) -> Integer.compare(a[0], b[0])); 
        PriorityQueue<Integer> pq = new PriorityQueue<>(); 
        pq.offer(intervals[0][1]); 
        for(int i = 1; i < intervals.length; i++){
            int nextStartTime = intervals[i][0]; 
            int nextEndTime = intervals[i][1]; 
            if(!pq.isEmpty() && pq.peek() <= nextStartTime) pq.poll(); 
            pq.offer(nextEndTime); 
        }
        return pq.size(); 
    }
}